tests/list-test.cc

```diff
@@ -0,0 +1,211 @@
+// Copyright Microsoft and CHERIoT Contributors.
+// SPDX-License-Identifier: MIT
+
+#define TEST_NAME "List"
+#include "tests.hh"
+#include <ds/linked_list.h>
+
+namespace
+{
+	using ObjectRing = ds::linked_list::cell::Pointer;
+
+	/**
+	 * Example class we want to link into a doubly linked list.
+	 *
+	 * The class contains a single integer for the purposes of the test.
+	 *
+	 * `ds::linked_list` is an intrusive list: we embed the list node into
+	 * the class we want to link.  There are various implementations of the
+	 * list nodes. Here we use the most simple one
+	 * (`ds::linked_list::cell::Pointer`) which relies on two pointers
+	 * `next` and `prev`.
+	 */
+	struct LinkedObject
+	{
+		int data;
+		/**
+		 * List node: links objects into the doubly-linked list.
+		 */
+		ObjectRing ring __attribute__((__cheri_no_subobject_bounds__)) = {};
+		/**
+		 * Container-of for the above field. This is used to retrieve the
+		 * corresponding object from a list element.
+		 */
+		__always_inline static struct LinkedObject *from_ring(ObjectRing *c)
+		{
+			return reinterpret_cast<struct LinkedObject *>(
+			  reinterpret_cast<uintptr_t>(c) -
+			  offsetof(struct LinkedObject, ring));
+		}
+	};
+} // namespace
+
+/**
+ * The sentinel is a special class encapsulating a list node. It provides many
+ * handy functions to operate on the list. The sentinel is thus part of the
+ * list, without being one of the objects linked.
```

Comment:
This is all true, but it might skip over some steps.  Perhaps something like...

`ds::linked_list`-s are circular, doubly-linked collections.  While they can stand on their own as rings of objects, it is sometimes convenient to create a designated "sentinel" node that participates in the collection without being part of the collection: it naturally provides pointers to the effective head and tail of the collection (the successor and predecessor of the sentinel, respectively) while not having to special-case "the collection is empty" in as many places as some other representations (that is, collections with sentinels need fewer "null pointer" checks).  Towards that end, the `Sentinel` class captures the notion of a designated not-an-element node and provides many handy functions to operate on the list.

Reply:
That's great! I'm adding it :slightly_smiling_face: